### PR TITLE
ICU-21311 Fix code generation in MeasureUnitTest.java and use it

### PR DESCRIPTION
**Pull request by hugovdm**
* I will update http://site.icu-project.org/design/formatting/measureformat/updating-measure-unit as part of the work of this PR.

##### Checklist

- [X] Issue filed: https://unicode-org.atlassian.net/browse/ICU-21311
- [X] Updated PR title and link in previous line to include Issue number
- [X] Issue accepted
- [ ] Tests included
- [ ] Documentation is changed or added

### CODE REVIEW

File: icu4c/source/i18n/unicode/measunit.h [hunk 1]
```diff
@@ -563,7 +563,7 @@ class U_I18N_API MeasureUnit: public UObject {
     /**
      * Returns by value, unit of acceleration: g-force.
      * Also see {@link #createGForce()}.
-     * @stable ICU 64
+     * @draft ICU 64
```
Comment (hugovdm) [inline icu4c/source/i18n/unicode/measunit.h:566]: Something is missing in the codebase, resulting in the "stable" annotations from dcfe6c4adae dropping back to "draft". Is it DRAFT_VERSIONS at the top of MeasureUnitTest.java that wasn't updated?

**Review by hugovdm — COMMENTED** [review]
Wish I could "request changes" on my own PR. :-) I want to mark it as "not ready for submit", but still "ready for review".

File: icu4c/source/i18n/measunit.cpp [hunk 2]
```diff
@@ -547,9 +517,7 @@ static const char * const gSubTypes[] = {
     "solar-mass",
     "stone",
     "ton",
-    "", // TODO(ICU-21076): manual edit of what should have been generated by Java.
-    "percent", // TODO(ICU-21076): regenerate, deal with duplication.
-    "permille", // TODO(ICU-21076): regenerate, deal with duplication.
```
Comment (hugovdm) [inline icu4c/source/i18n/measunit.cpp:552]: These fall under "concentr". I expect they were added under "none" when we didn't have the "concentr" type yet.

File: icu4c/source/i18n/measunit.cpp [hunk 3]
```diff
@@ -55,44 +54,15 @@ static const int32_t gOffsets[] = {
...
-    442,
-    446,
+    424,
+    430,
+    440,
+    444,
+    448,
     450,
-    452,
-    486
```
Comment (hugovdm) [inline icu4c/source/i18n/measunit.cpp:64]: Renumbering due to removal of duplicate "percent" and "permille" under the "none" category.

File: icu4c/source/i18n/measunit.cpp [hunk 4]
```diff
@@ -2271,9 +2237,11 @@ StringEnumeration* MeasureUnit::getAvailableTypes(UErrorCode &errorCode) {
 }
 
 bool MeasureUnit::findBySubType(StringPiece subType, MeasureUnit* output) {
+    // Sanity checking kCurrencyOffset and final entry in gOffsets
+    U_ASSERT(uprv_strcmp(gTypes[kCurrencyOffset], "currency") == 0);
+    U_ASSERT(gOffsets[UPRV_LENGTHOF(gOffsets) - 1] == UPRV_LENGTHOF(gSubTypes));
+
     for (int32_t t = 0; t < UPRV_LENGTHOF(gOffsets) - 1; t++) {
-        // Ensure kCurrencyOffset is set correctly
-        U_ASSERT(uprv_strcmp(gTypes[kCurrencyOffset], "currency") == 0);
```
Comment (hugovdm) [inline icu4c/source/i18n/measunit.cpp:2276]: I guess it was from pure inattentiveness that I placed this assert inside the for loop before.
I feel this assert is partially documentation for how the "skip currency units" code works, so it fits better for me than the assert I'm adding in this PR:

It doesn't feel like this function is the ideal place for the new assert, but I don't find a "more ideal" place. (In Java, a "static intialization block" might have served?)

File: icu4j/main/tests/core/src/com/ibm/icu/dev/test/format/MeasureUnitTest.java [hunk 5]
```diff
@@ -2898,6 +2898,7 @@ public void test20332_PersonUnits() {
     // for MeasureFormat during the release process.
     static void generateCXXHConstants(String thisVersion) {
```
Comment (hugovdm) [inline icu4j/main/tests/core/src/com/ibm/icu/dev/test/format/MeasureUnitTest.java:2899]: I have added the start and end comments to the generated output code. I was looking to make it simpler to find where the output code fits. (It enabled generating all the generated code blocks at the same time - which I then saw isn't the most comfortable workflow, but it seems nice for it to be possible.)

File: icu4j/main/tests/core/src/com/ibm/icu/dev/test/format/MeasureUnitTest.java [hunk 6]
```diff
@@ -3085,27 +3094,6 @@ static void generateCXXConstants() {
                     measureUnitToTypeSubType.get(entry.getKey()));
         }
 
-        System.out.println("// Must be sorted by first value and then second value.");
-        System.out.println("static int32_t unitPerUnitToSingleUnit[][4] = {");
```
Comment (hugovdm) [inline icu4j/main/tests/core/src/com/ibm/icu/dev/test/format/MeasureUnitTest.java:3089]: Deleting this block implements the TODO comment wherein the output was manually removed:

// unitPerUnitToSingleUnit no longer in use! TODO: remove from code-generation code.

File: icu4j/main/tests/core/src/com/ibm/icu/dev/test/format/MeasureUnitTest.java [hunk 7]
```diff
@@ -3286,16 +3276,7 @@ else if (isDraft(name)) {
                 System.out.println();
             }
         }
-        System.out.println("    private static HashMap<Pair<MeasureUnit, MeasureUnit>, MeasureUnit>unitPerUnitToSingleUnit =");
```
Comment (hugovdm) [inline icu4j/main/tests/core/src/com/ibm/icu/dev/test/format/MeasureUnitTest.java:3289]: As above, unitPerUnitToSingleUnit is no longer in use / no longer exists.

**Review by sffc — DISMISSED** [review]
LGTM; wait a bit for Peter's review before meting.

**Review by pedberg-icu — APPROVED** [review]
Looks good, thanks, this will save some manual touchup during CLDR-ICU integration.

**Comment by hugovdm** [comment]
Moving to the master branch. I can't make the force-push be clean: I need to drop commits that are in the maint branch but not in the master branch.

**Review by pedberg-icu — APPROVED** [review]

**Comment by hugovdm** [comment]
https://sites.google.com/site/icusite/design/formatting/measureformat/updating-measure-unit updated.

**Pull request merged**